src/serac/physics/tests/serac_thermal_mechanics_functional.cpp

```diff
@@ -0,0 +1,397 @@
+                                                         .max_iter    = 500,
+                                                         .lin_solver  = LinearSolver::GMRES,
+                                                         .prec        = HypreBoomerAMGPrec{}};
+
+  const NonlinearSolverOptions default_nonlinear_options = {
+      .rel_tol = 1.0e-4, .abs_tol = 1.0e-8, .max_iter = 10, .print_level = 1};
+
+  const SolverOptions solid_options = {default_linear_options, default_nonlinear_options};
+
+  ThermalMechanicsFunctional<p, dim> thermal_solid_solver(thermal_options, solid_options, GeometricNonlinearities::On,
+                                                          "thermal_solid_functional");
+
+  double                                       rho       = 1.0;
+  double                                       E         = 1.0;
+  double                                       nu        = 0.0;
+  double                                       c         = 1.0;
+  double                                       alpha     = 1.0e-3;
+  double                                       theta_ref = 2.0;
+  double                                       k         = 1.0;
+  GreenSaintVenantThermoelasticMaterial        material{rho, E, nu, c, alpha, theta_ref, k};
+  GreenSaintVenantThermoelasticMaterial::State initial_state{};
+  auto                                         qdata = thermal_solid_solver.createQuadratureDataBuffer(initial_state);
+  thermal_solid_solver.setMaterial(material, qdata);
+
+  // Define the function for the initial temperature
+  double theta_0                   = 1.0;
+  auto   initial_temperature_field = [theta_0](const mfem::Vector&, double) -> double { return theta_0; };
+
+  auto one = [](const mfem::Vector&, double) -> double { return 1.0; };
+
+  // Set the initial temperature and boundary condition
+  // thermal_solid_solver.setTemperatureBCs(ess_bdr, theta_0);
+  thermal_solid_solver.setTemperatureBCs(temp_bdr, one);
+  thermal_solid_solver.setTemperature(initial_temperature_field);
+
+  // Define the function for the disolacement boundary condition
+  auto zeroVector = [](const mfem::Vector&, mfem::Vector& u) { u = 0.0; };
+
+  // Set the initial displacement and boundary condition
+  thermal_solid_solver.setDisplacementBCs(constraint_bdr, zeroVector);
+  thermal_solid_solver.setDisplacement(zeroVector);
+
+  // Finalize the data structures
+  thermal_solid_solver.completeSetup();
+
+  // thermal_solid_solver.initializeOutput(serac::OutputType::VisIt, "thermal_mechanics_without_input_file_output");
+
+  // dump initial state to output
+  // thermal_solid_solver.outputState();
+
+  // Perform the quasi-static solve
+  double dt = 1.0;
+  thermal_solid_solver.advanceTimestep(dt);
+
+  // thermal_solid_solver.outputState();
+
+  // Check the final displacement norm
+  EXPECT_NEAR(expected_norm, norm(thermal_solid_solver.displacement()), 1.0e-4);
+}
+
+// TODO: investigate this failing test
```

Comment:
Is this still failing?

Reply:
it is failing because there was a placeholder analytic solution that isn't quite what we want. In general though we don't (or at least I don't) know a good way to check for the "right" answer. The actual analytic solution to this problem will never be found by the thermomechanics module as-written since the operator-split approach doesn't actually enforce that mechanical and thermal equilibrium are simultaneously satisfied.

One possibility might be to phrase this as a time-dependent problem (applying the deformation gradually), and then it should approximate the "right" answer in the limit, as you take enough time-steps. That would make the test much more expensive though, I'm not sure what the right thing is to do with this test.

Reply:
Sam and I worked on these together, and I voted to disable this test, for the reasons Sam gave above. I have a suggestion to replace it: we could manufacture a solution for the coupled problem, set the discrete temperature and displacements fields to that solution, and check that both residuals are zero. It's not a foolproof test, but it's not trivial, either. This will be easier to do once the patch test/manufactured solution PR comes through.

Reply:
Looking over the other two test: they pass, but don't really do anything (the solutions are practically trivial). I wonder: is it worth stealing some of the work from the patch-test branch to make them into more meaningful tests, or should we just try to get this PR in first and fix them in the patch-test PR?

Reply:
I'd prefer the latter so that we can see if patch testing plain mechanics/heat transfer reveals any issues, and also so that I can participate on the thermomechanics testing.

Reply:
Agreed. This can wait for the patch test PR. I'd remove it here to avoid confusion.